nibabel/streamlines/tractogram.py

```diff
@@ -0,0 +1,677 @@
+
+            data_per_point_keys = first_item.data_for_points.keys()
+            for k in data_per_point_keys:
+                lazy_tractogram._data_per_point[k] = _gen(k)
+
+        except StopIteration:
+            pass
+
+        return lazy_tractogram
+
+    @property
+    def streamlines(self):
+        streamlines_gen = iter([])
+        if self._streamlines is not None:
+            streamlines_gen = self._streamlines()
+        elif self._data is not None:
+            streamlines_gen = (t.streamline for t in self._data())
+
+        # Check if we need to apply an affine.
+        if not np.all(self._affine_to_apply == np.eye(4)):
+            def _apply_affine():
+                for s in streamlines_gen:
+                    yield apply_affine(self._affine_to_apply, s)
+
+            return _apply_affine()
+
+        return streamlines_gen
+
+    @streamlines.setter
+    def streamlines(self, value):
+        if value is not None and not callable(value):
+            raise TypeError("`streamlines` must be a generator function.")
+
+        self._streamlines = value
+
+    @property
+    def data_per_streamline(self):
+        return self._data_per_streamline
+
+    @data_per_streamline.setter
+    def data_per_streamline(self, value):
+        self._data_per_streamline = LazyDict(self, value)
+
+    @property
+    def data_per_point(self):
+        return self._data_per_point
+
+    @data_per_point.setter
+    def data_per_point(self, value):
+        self._data_per_point = LazyDict(self, value)
+
+    @property
+    def data(self):
+        if self._data is not None:
+            return self._data()
+
+        def _gen_data():
+            data_per_streamline_generators = {}
+            for k, v in self.data_per_streamline.items():
+                data_per_streamline_generators[k] = iter(v)
+
+            data_per_point_generators = {}
+            for k, v in self.data_per_point.items():
+                data_per_point_generators[k] = iter(v)
+
+            for s in self.streamlines:
+                data_for_streamline = {}
+                for k, v in data_per_streamline_generators.items():
+                    data_for_streamline[k] = next(v)
+
+                data_for_points = {}
+                for k, v in data_per_point_generators.items():
+                    data_for_points[k] = next(v)
+
+                yield TractogramItem(s, data_for_streamline, data_for_points)
+
+        return _gen_data()
+
+    def __getitem__(self, idx):
+        raise NotImplementedError('`LazyTractogram` does not support indexing.')
+
+    def __iter__(self):
+        i = 0
+        for i, tractogram_item in enumerate(self.data, start=1):
+            yield tractogram_item
+
+        # Keep how many streamlines there are in this tractogram.
+        self._nb_streamlines = i
+
+    def __len__(self):
+        # Check if we know how many streamlines there are.
+        if self._nb_streamlines is None:
```

Comment:
Seems like this could be pretty confusing.  I guess it means that if you ask for the length, and the object has not yet been iterated over, the user gets the length, but the object (certainly?  perhaps?) becomes useless - in fact deceptive, because presumably before `len(obj)` you might be able to get 100 tracks out of the object, but after `len(obj)` you get 0 tracks out of the object.

So - raise an error rather than warn if the length is not available?